# Advanced Pandas Assignment 

### Q1 Dataset Exploration
### Perform the following exploratory steps:
1. Display the first 10 rows
2. Display the dataset structure using info()
3. Find the number of rows and columns
4. Identify the data types of each column

#### Answer:

In [1]:
# Importing the Pandas library.
import pandas as pd

In [2]:
# Loading the dataset.
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

In [8]:
# 1. Displaying first 10 rows
print("First 10 Rows:")
df.head(10)

First 10 Rows:


,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [9]:
# 2. Dataset structure
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


In [10]:
# 3. Number of rows and columns
print("Shape of dataset (rows, columns):")
print(df.shape)

Shape of dataset (rows, columns):
(230, 11)


In [12]:
# 4. Data types of each column
print(df.dtypes)

Order_ID        int64
Order_Date     object
Customer       object
City           object
Region         object
Product        object
Category       object
Sales           int64
Quantity        int64
Discount      float64
Profit        float64
dtype: object


### Q2 Sales Filtering Analysis
### Filter the dataset to display:
1. Orders where Sales > 3500
2. Orders where Region = "East" and Sales > 2500
### Explain what insights you observe.

#### Answer:

In [19]:
# 1. Orders where Sales > 3500

high_sales = df[df['Sales'] > 3500]
high_sales

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
10,1011,2024-06-30,Neha,Delhi,South,Printer,Office,5977,6,0.00,1578.32
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [18]:
# 2. Orders where Region = "East" and Sales > 2500

east_high_sales = df[(df['Region'] == 'East') & (df['Sales'] > 2500)]
east_high_sales

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
34,1035,2024-02-20,Nitin,Ahmedabad,East,Laptop,Electronics,2535,4,0.10,188.15
41,1042,2024-07-15,Rohan,Chennai,East,Laptop,Electronics,3268,8,0.15,295.15
43,1044,2024-09-20,Kavya,Hyderabad,East,Printer,Office,2618,7,0.05,454.64


In [21]:
# For batter insights

print("Most repeated high value category overall:")
print(high_sales['Category'].value_counts())

print("Top high value  Category overall:")
print(high_sales['Category'].value_counts().idxmax())

print("Most repeated high value category in east region:")
print(east_high_sales['Category'].value_counts())

print("Top high value Category in east region:")
print(east_high_sales['Category'].value_counts().idxmax())

Most repeated high value category overall:
Category
Electronics    56
Accessories    39
Office          9
Name: count, dtype: int64
Top high value  Category overall:
Electronics
Most repeated high value category in east region:
Category
Electronics    34
Accessories    11
Office          3
Name: count, dtype: int64
Top high value Category in east region:
Electronics


#### Insights:
- The most repeated high-value category overall is Electronics (56 orders), followed by Accessories and Office products.
  
- In the East region, Electronics again dominates with 34 high-value orders, significantly higher than other categories. This shows that the East region has a particularly strong market for Electronics compared to Accessories and Office supplies.

- Electronics is the top-performing category both overall and in the East region

### Q3 Customer Purchase Analysis
### Identify the top 5 customers based on total sales.
(Hint: Use GroupBy with aggregation.)

#### Answer:
##### Steps:
1. Group the dataset by Customer Name
2. Calculate total sales using sum()
3. Sort the values in descending order
4. Select the top 5 customers

In [27]:
# Grouping Customers and calculateing total sales
customer_sales = df.groupby('Customer')['Sales'].sum()


In [31]:
# Sorting in descending order and printing top 5 customers.
top_5_customers = customer_sales.sort_values(ascending=False).head(5)
print(top_5_customers)

Customer
Nitin    59820
Raj      58416
Rahul    52291
Dev      48500
Isha     46818
Name: Sales, dtype: int64


### Q4 Regional Sales Performance
### Calculate the total sales generated by each region.
### Then identify:
- Which region generated the highest revenue?

#### Answer:
##### Steps:
1. Group the dataset by 'Region'
2. Calculate total sales using sum()
3. Compare values to identify the highest-performing region



In [34]:
# Calculateing total sales by region.
region_sales = df.groupby('Region')['Sales'].sum()
print("Total Sales by Region:")
print(region_sales)

Total Sales by Region:
Region
East     232675
North    158210
South    183108
West     170227
Name: Sales, dtype: int64


In [35]:
# Region with highest revenue.
top_region = region_sales.idxmax()
print("Region with Highest Revenue:")
print(top_region)

Region with Highest Revenue:
East


### Q5 Product Performance Analysis
### Calculate the total quantity sold for each product. And Which product appears to be the most popular?

#### Answer:
##### Steps:
1. Group the dataset by 'Product' and Calculate total quantity using sum().
2. Identify the product with the highest quantity sold


In [38]:
# Calculate total quantity sold for each product
product_quantity = df.groupby('Product')['Quantity'].sum()
print(product_quantity)

Product
Headphones    115
Keyboard      110
Laptop        140
Monitor       156
Mouse         122
Phone         151
Printer       108
Tablet        150
Name: Quantity, dtype: int64


In [37]:
# Most popular product
most_popular_product = product_quantity.idxmax()
print("Most Popular Product:")
print(most_popular_product)

Most Popular Product:
Monitor


### Q6 Multi-Level Grouping
- Group the dataset by: Region and Category
- Calculate the total sales for each combination.
- What insights can you derive from this analysis?

#### Answers:
##### Steps:
1. Group the dataset by both 'Region' and 'Category' and Calculate total sales using sum().
2. Sum all categories within each region.
3. Insights

In [39]:
# Step 1 :
region_category_sales = df.groupby(['Region', 'Category'])['Sales'].sum()
print(region_category_sales)

Region  Category   
East    Accessories     57646
        Electronics    160907
        Office          14122
North   Accessories     62052
        Electronics     70023
        Office          26135
South   Accessories     80914
        Electronics     86690
        Office          15504
West    Accessories     64811
        Electronics     84128
        Office          21288
Name: Sales, dtype: int64


In [41]:
# Step 2 :
region_total_sales = region_category_sales.groupby('Region').sum()
print("Total Sales for Each Region:")
print(region_total_sales)

Total Sales for Each Region:
Region
East     232675
North    158210
South    183108
West     170227
Name: Sales, dtype: int64


#### Insights:
- The East region generates the highest total sales (232,675), making it the top-performing region overall.
- Across all regions, Electronics is the highest revenue-generating category.
- Office category contributes the least in every region, indicating lower sales volume or demand.


#### Conclusion:
  - Focus on Electronics for maximum revenue.
  - Strengthen strategies in the North region.
  - Maintain strong performance in the East region, especially in Electronics.

### Q7 Pivot Table Creation
### Create a pivot table showing:
1. Rows → Region
2. Columns → Product
3. Values → Total Sales
- Which region sells the most laptops?



#### Answer:

In [42]:
# Createing pivot table.

pivot_table = pd.pivot_table(
    df,
    values='Sales',
    index='Region',
    columns='Product',
    aggfunc='sum'
)
print(pivot_table)

Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          23899     10845   55568    31299  22902  39667    14122   34373
North         15450     20538    4540    14922  26064  30535    26135   20026
South         24111     23480   23029    19167  33323  18218    15504   26276
West          17159     38499   10826    18178   9153  27850    21288   27274


In [43]:
# Find which region sells the most laptops
top_laptop_region = pivot_table['Laptop'].idxmax()
print("Region with Highest Laptop Sales:")
print(top_laptop_region)

Region with Highest Laptop Sales:
East


### Q8 Profitability Analysis
### Calculate the average profit for each product category. And Which category is most profitable?

#### Answers:
##### Steps-
1. Group the dataset by 'Category' and Calculate the average profit using ```.mean()```
2. Identify the most profitable category using ```.idxmax()```

In [50]:
# Step 1 :
category_profit = df.groupby('Category')['Profit'].mean()
print("Average Profit by Category:")
print(category_profit)

Average Profit by Category:
Category
Accessories    410.603086
Electronics    396.462114
Office         471.938462
Name: Profit, dtype: float64


In [51]:
# Step 1 :
most_profitable_category = category_profit.idxmax()
print("Most Profitable Category:")
print(most_profitable_category)

Most Profitable Category:
Office


### Q9 Discount Impact Study
### Analyze the relationship between discount and profit.
#### Tasks:
1. Calculate the average profit for each discount level.
2. Identify whether higher discounts lead to lower profit.

#### Answer:

In [53]:
# Group the dataset by 'Discount' and Calculate the average profit.
discount_profit = df.groupby('Discount')['Profit'].mean()
print("Average Profit for Each Discount Level:")
print(discount_profit)

Average Profit for Each Discount Level:
Discount
0.00    751.210588
0.05    531.368039
0.10    346.460513
0.15    237.191400
0.20     90.026923
Name: Profit, dtype: float64


In [54]:
# Sorting data for better understanding
print("Sorted (by Discount Level):")
print(discount_profit.sort_index())

Sorted (by Discount Level):
Discount
0.00    751.210588
0.05    531.368039
0.10    346.460513
0.15    237.191400
0.20     90.026923
Name: Profit, dtype: float64


#### Insights:
- There is a **clear negative relationship between discount and profit**.
- As discount increases, average profit consistently decreases:
- This shows that higher discounts reduce profit margins.

### Q10 End-to-End Data Wrangling Task
### Perform the following workflow:
1. Filter the dataset where Sales > 3000.
2. Group data by Region and Product.
3. Calculate total sales.
4. Reshape the result into a pivot table.
- Explain the business insights obtained from this analysis.

#### Answer:
##### Steps-
1. Filter data where Sales > 3000
2. Group by Region and Product and calculate total sales
4. Reshape the data into a pivot table

In [55]:
# Step 1:
high_value_sales = df[df['Sales'] > 3000]

In [56]:
# Step 2:
grouped_data = high_value_sales.groupby(['Region', 'Product'])['Sales'].sum()

In [58]:
# Step 3:
pivot_table = grouped_data.unstack()
print(pivot_table)

Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          14469      9642   44096    16420  19000  29500     3696   29229
North          5326     15653    4540     5087  22209  24670    17185    3392
South         10026     19249   17950    14921  26928  18218    14143   19613
West          13898     33380    5922    14354   3511  15653    12648   17185


#### Insights:
- Laptops are highly concentrated in the East region, making it the premium market
- Accessories (Mouse, Keyboard, Headphones) perform well across all regions
- Product preference varies by region, indicating need for region-specific strategy

#### Business Value:
- Focus **premium products (Laptops, Phones)** in East
- Promote **balanced product mix** in South
- Target **accessories market** in West
- Improve **high-value product penetration** in North
